### PR TITLE
feat(48665): tsconfig "extends" field ignores "exports" field of source package 

### LINKED ISSUE #48665: tsconfig "extends" field ignores "exports" field of source package
# Bug Report

### 🔎 Search Terms

tsconfig, exports, extends

### 🕗 Version & Regression Information

I encountered this issue in version `typescript@4.7.0-dev.20220408`.

### Error description

You can use the tsconfig.json [extends](https://www.typescriptlang.org/tsconfig#extends) field to inherit configuration options from another tsconfig file. The base tsconfig file can be located in another package, like demonstrated [here](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#tsconfig-bases).

### 🙁 Actual behavior

Let's say I'm in package `my-app`, which contains a tsconfig with `"extends": "my-lib/tsconfig-base.json`. Typescript ignores the [exports](https://nodejs.org/api/packages.html#exports) field of package `my-lib`. It will always look for a file `tsconfig-base.json` in the my-lib root directory, even if an "exports" mapping exists for this name.

### 🙂 Expected behavior

I expected Typescript to use the standard node resolution algorithm when looking up the tsconfig file on disk, which respects the "exports" field.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #48665

### CODE REVIEW

**Comment by a-tarasyuk** [comment]
> it should just need to be swapped to nodeNextModuleNameResolver and pass { moduleResolution: ModuleResolutionKind.NodeNext }

I tried however it didn't process `tsconfig` files, I added changes to `loadJSOrExactTSFileName`. Maybe need to add additional tests to cover more cases. @weswigham Could you please review these changes?

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -1868,14 +1873,19 @@ function loadModuleFromFileNoImplicitExtensions(extensions: Extensions, candidat
     }
 }
 
-function loadJSOrExactTSFileName(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
+function loadFileName(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
```
Comment (andrewbranch) [inline src/compiler/moduleNameResolver.ts:1876]: Yeah, this name was already bad even before adding a JSON config file lookup to it. I tried to come up with a better one in my refactors but gave up. The distinctive thing about this function is it’s only ever called with paths written in package.json files—never module specifiers written in source files—and so consequently it always allows the candidate to end with a TS extension (but will also try substituting a JS extension for a TS extension). I guess my best suggestion is

```suggestion
/**
 * This function is it’s only ever called with paths written in package.json files - never
 * module specifiers written in source files - and so consequently it always allows the
 * candidate to end with a TS extension (but will also try substituting a JS extension for
 * a TS extension).
 */
function loadFileNameFromPackageJsonField(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
```

but maybe @sandersn has naming advice as well.

**Review by andrewbranch — APPROVED** [review]
Implementation looks good, but would like to get more naming ideas

File: src/compiler/moduleNameResolver.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/compiler/moduleNameResolver.ts]: Thanks for the great suggestion 👍🏻 , I was thinking about the name of this function and decided to use the general name `loadFileName` :).

File: src/compiler/moduleNameResolver.ts [hunk 2]
```diff
@@ -1868,14 +1873,24 @@ function loadModuleFromFileNoImplicitExtensions(extensions: Extensions, candidat
     }
 }
 
-function loadJSOrExactTSFileName(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
+/**
+ * This function is it’s only ever called with paths written in package.json files - never
```
Comment (sandersn) [inline src/compiler/moduleNameResolver.ts:1877]:
```suggestion
 * This function is only ever called with paths written in package.json files - never
```

File: src/compiler/moduleNameResolver.ts [hunk 3]
```diff
@@ -1868,14 +1873,24 @@ function loadModuleFromFileNoImplicitExtensions(extensions: Extensions, candidat
     }
 }
 
-function loadJSOrExactTSFileName(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
+/**
+ * This function is it’s only ever called with paths written in package.json files - never
+ * module specifiers written in source files - and so consequently it always allows the
```
Comment (sandersn) [inline src/compiler/moduleNameResolver.ts:1878]:
```suggestion
 * module specifiers written in source files - and so it always allows the

```

File: src/compiler/moduleNameResolver.ts [hunk 4]
```diff
@@ -1868,14 +1873,24 @@ function loadModuleFromFileNoImplicitExtensions(extensions: Extensions, candidat
     }
 }
 
-function loadJSOrExactTSFileName(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
+/**
+ * This function is it’s only ever called with paths written in package.json files - never
+ * module specifiers written in source files - and so consequently it always allows the
+ * candidate to end with a TS extension (but will also try substituting a JS extension for a TS extension).
+ */
+function loadFileNameFromPackageJsonField(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
```
Comment (sandersn) [inline src/compiler/moduleNameResolver.ts:1881]: the two preceding functions start with `loadModuleFrom*`, but `candidate` isn't really a module specifier, so I don't know. I think the current suggestion is the best, but maybe `loadModuleFromPackageJsonField` isn't *too* inaccurate either. And it's parallel to the earlier functions.

(I'd probably stick with what you have, though, unless you like my suggestion a *lot* better.)

**Review by sandersn — COMMENTED** [review]
No real new suggestions, but a couple of edits to the English.

File: src/compiler/moduleNameResolver.ts (hunk 4 above)
Comment (andrewbranch) [reply src/compiler/moduleNameResolver.ts:1881]: It is definitely more filenamey than not.

**Pull request merged**